### ISSUE TITLE
lerna 3 didn't throw Error when npm publish failed

### ISSUE THREAD

**Issue by tweet** [lerna/lerna#1873]
We use lerna 3 to manage a multi-project repo, and command `yarn lerna publish` as one step in build & release process. lerna 3 didn't throw Error when npm publish failed, please see the log attached below.
## Expected Behavior

**Lerna's process should throw Error**, then the build process can end or ignore the following steps.
## Current Behavior

Lerna only log some error message when execute npm publish failed, but not exit with error.
## Steps to Reproduce (for bugs)

1. Change your npm publishConfig's registry to a wrong url.
2. Execute lerna publish command.
3. Check the result.

Lerna can throw Error when other errors thrown, like git error:
<details><summary>lerna-debug.log</summary><p>

```json
2019-01-16T06:59:55.3966330Z lerna ERR! lerna Command failed: git remote update
2019-01-16T06:59:55.3984960Z lerna ERR! lerna fatal: could not read Username for 'http://internal.company.com:8080': No such device or address
2019-01-16T06:59:55.4003480Z lerna ERR! lerna error: Could not fetch origin
2019-01-16T06:59:55.4021370Z lerna ERR! lerna 
2019-01-16T06:59:55.4038410Z lerna ERR! lerna Fetching origin
2019-01-16T06:59:55.4055930Z lerna ERR! lerna 
2019-01-16T06:59:55.4073000Z error Command failed with exit code 1.
2019-01-16T06:59:55.4091100Z info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
2019-01-16T06:59:55.4256660Z ##[error]Error: /usr/local/nodejs/bin/yarn failed with return code: 1
```
</p></details>

lerna 3 didn't throw Error when npm publish failed.
<details><summary>lerna-debug.log</summary><p>

```txt
2019-01-16T07:16:34.3626500Z lerna info publish Publishing packages to npm...
2019-01-16T07:16:34.3646140Z lerna notice Skipping all user and access validation due to third-party registry
2019-01-16T07:16:34.3665700Z lerna notice Make sure you're authenticated properly ¯\_(ツ)_/¯
2019-01-16T07:16:34.3686340Z lerna WARN ENOLICENSE Packages @lerna-repo/repo1, @lerna-repo/repo2 are missing a license
2019-01-16T07:16:34.3707520Z lerna sill getCurrentSHA 
2019-01-16T07:16:34.3726980Z lerna verb getCurrentSHA 394d73d06d13a133729d1704141997a8e069706c
2019-01-16T07:16:34.3773140Z lerna sill lifecycle No script for "prepublish" in "root", continuing
2019-01-16T07:16:34.3792790Z lerna sill lifecycle No script for "prepare" in "root", continuing
2019-01-16T07:16:34.3811160Z lerna sill lifecycle No script for "prepublishOnly" in "root", continuing
2019-01-16T07:16:34.3830480Z lerna sill lifecycle No script for "prepack" in "root", continuing
2019-01-16T07:16:34.3849400Z lerna verb pack-directory packages/repo1
2019-01-16T07:16:34.3867480Z lerna verb pack-directory packages/repo2
2019-01-16T07:16:34.3886680Z lerna sill lifecycle No script for "prepublish" in "@lerna-repo/repo1", continuing
2019-01-16T07:16:34.3906040Z lerna sill lifecycle No script for "prepublish" in "@lerna-repo/repo2", continuing
2019-01-16T07:16:34.3925570Z lerna sill lifecycle No script for "prepare" in "@lerna-repo/repo1", continuing
2019-01-16T07:16:34.3945480Z lerna sill lifecycle No script for "prepare" in "@lerna-repo/repo2", continuing
2019-01-16T07:16:34.3966790Z lerna sill lifecycle No script for "prepublishOnly" in "@lerna-repo/repo1", continuing
2019-01-16T07:16:34.3986380Z lerna sill lifecycle No script for "prepublishOnly" in "@lerna-repo/repo2", continuing
2019-01-16T07:16:34.4005430Z lerna sill lifecycle No script for "prepack" in "@lerna-repo/repo1", continuing
2019-01-16T07:16:34.4024170Z lerna sill lifecycle No script for "prepack" in "@lerna-repo/repo2", continuing
2019-01-16T07:16:34.4272240Z lerna sill lifecycle No script for "postpack" in "@lerna-repo/repo1", continuing
2019-01-16T07:16:34.4291390Z lerna verb packed @lerna-repo/repo1 packages/repo1
2019-01-16T07:16:34.4312180Z lerna sill lifecycle No script for "postpack" in "@lerna-repo/repo2", continuing
2019-01-16T07:16:34.4331660Z lerna verb packed @lerna-repo/repo2 packages/repo2
2019-01-16T07:16:34.4350450Z lerna sill lifecycle No script for "postpack" in "root", continuing
2019-01-16T07:16:34.4370010Z lerna verb publish @lerna-repo/repo1
2019-01-16T07:16:34.4392410Z lerna verb publish @lerna-repo/repo2
2019-01-16T07:16:34.4762960Z lerna http fetch PUT 404 http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2 25ms
2019-01-16T07:16:34.4797080Z lerna ERR! Error: 404 Not Found - PUT http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2
2019-01-16T07:16:34.4823670Z lerna ERR!     at res.buffer.catch.then.body (node_modules/npm-registry-fetch/check-response.js:104:15)
2019-01-16T07:16:34.4842770Z lerna ERR!     at <anonymous>
2019-01-16T07:16:34.4862020Z lerna ERR! lerna 404 Not Found - PUT http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2
2019-01-16T07:16:34.4917030Z Done in 3.40s.
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.10.5|
| `npm --version`  | 6.4.1|
| `yarn --version` | 1.12.3|
| `node --version` | 10.10.0|

| OS | Version |
| --- | --- |
| Windows 10 | 1809|

**Comment by evocateur** [comment]
Without the _entire_ `lerna-debug.log` in addition to the _exact_ scripts you called, I can't begin to diagnose. A wild guess, probably forgot to set `process.exitCode` somewhere in a `.catch()` handler.

**Comment by tweet** [comment]
The entire `lerna-debug.log` is here.
<details><summary>lerna-debug.log</summary><p>


```txt
0 silly argv { _: [ 'publish' ],
0 silly argv   yes: true,
0 silly argv   y: true,
0 silly argv   composed: 'publish',
0 silly argv   lernaVersion: '3.10.5',
0 silly argv   '$0': 'node_modules\\lerna\\cli.js' }
1 notice cli v3.10.5
2 verbose rootPath D:\lerna-repo
3 info versioning independent
4 verbose session 5144b5684a90036b
5 verbose user-agent lerna/3.10.5/node@v10.10.0+x64 (win32)
6 verbose git-describe undefined => "@lerna-repo/repo2@2.1.1-3-g523e8d1"
7 silly git-describe parsed => {"lastTagName":"@lerna-repo/repo2@2.1.1","lastVersion":"2.1.1","refCount":"3","sha":"523e8d1","isDirty":false}
8 silly argv { _: [ 'publish' ],
8 silly argv   yes: true,
8 silly argv   y: true,
8 silly argv   composed: 'publish',
8 silly argv   lernaVersion: '3.10.5' }
9 verbose rootPath D:\lerna-repo
10 silly isAnythingCommitted
11 verbose isAnythingCommitted 1
12 silly currentBranch
13 verbose currentBranch feat/test
14 silly remoteBranchExists
15 silly isBehindUpstream
16 silly isBehindUpstream feat/test is behind origin/feat/test by 0 commit(s) and ahead by 0
17 silly hasTags
18 verbose hasTags true
19 silly git-describe.sync "@lerna-repo/repo2@2.1.1-3-g523e8d1" => {"lastTagName":"@lerna-repo/repo2@2.1.1","lastVersion":"2.1.1","refCount":"3","sha":"523e8d1","isDirty":false}
20 info Looking for changed packages since @lerna-repo/repo2@2.1.1
21 silly diffSinceIn @lerna-repo/repo2@2.1.1 packages/repo1
22 silly diff
23 silly diffSinceIn @lerna-repo/repo2@2.1.1 packages/repo2
24 silly diff packages/repo2/CHANGELOG.md
24 silly diff packages/repo2/index.js
24 silly diff packages/repo2/package.json
25 verbose updated @lerna-repo/repo2
26 verbose git-describe undefined => "@lerna-repo/repo2@2.1.1-3-g523e8d1"
27 silly git-describe parsed => {"lastTagName":"@lerna-repo/repo2@2.1.1","lastVersion":"2.1.1","refCount":"3","sha":"523e8d1","isDirty":false}
28 silly independent for @lerna-repo/repo2 at D:\lerna-repo\packages\repo2
29 verbose independent increment 2.1.2 by patch
30 silly batched [ PackageGraphNode {
30 silly batched     name: '@lerna-repo/repo2',
30 silly batched     externalDependencies:
30 silly batched      Map {
30 silly batched        'babel-eslint' => [Object],
30 silly batched        'eslint' => [Object],
30 silly batched        'eslint-config-hwit' => [Object],
30 silly batched        'eslint-plugin-vue' => [Object],
30 silly batched        'lint-staged' => [Object],
30 silly batched        'vue-eslint-parser' => [Object] },
30 silly batched     localDependencies: Map {},
30 silly batched     localDependents: Map {} } ]
31 info auto-confirmed
32 silly lifecycle No script for "preversion" in "root", continuing
33 silly lifecycle No script for "preversion" in "@lerna-repo/repo2", continuing
34 silly lifecycle No script for "version" in "@lerna-repo/repo2", continuing
35 silly independent for @lerna-repo/repo2 at D:\lerna-repo\packages\repo2
36 silly independent writing new entry: "## [2.1.3](http://ifoundation.btit.huawei.com:8080/tfs/TFSCollection/ERP-MTS_MiddleTier/_git/lerna-repo/compare/@lerna-repo/repo2@2.1.1...@lerna-repo/repo2@2.1.3) (2019-01-17)\n\n**Note:** Version bump only for package @lerna-repo/repo2\n\n\n\n"
37 verbose independent wrote D:\lerna-repo\packages\repo2\CHANGELOG.md
38 silly lifecycle No script for "version" in "root", continuing
39 silly gitAdd [ 'D:\\lerna-repo\\packages\\repo2\\package.json',
39 silly gitAdd   'D:\\lerna-repo\\packages\\repo2\\CHANGELOG.md' ]
40 silly gitCommit chore(release): publish [ci skip] ***NO_CI***
40 silly gitCommit
40 silly gitCommit  - @lerna-repo/repo2@2.1.3
41 verbose git [ 'commit',
41 verbose git   '-F',
41 verbose git   'C:\\Users\\Test~1\\AppData\\Local\\Temp\\3450cd-c497-38r6-a95f-0bfca887b2c1\\lerna-commit.txt' ]
42 silly gitTag @lerna-repo/repo2@2.1.3
43 verbose git [ 'tag',
43 verbose git   '@lerna-repo/repo2@2.1.3',
43 verbose git   '-m',
43 verbose git   '@lerna-repo/repo2@2.1.3' ]
44 silly lifecycle No script for "postversion" in "@lerna-repo/repo2", continuing
45 silly lifecycle No script for "postversion" in "root", continuing
46 info git Pushing tags...
47 silly gitPush origin feat/test
48 silly batched [ PackageGraphNode {
48 silly batched     name: '@lerna-repo/repo2',
48 silly batched     externalDependencies: Map {},
48 silly batched     localDependencies: Map {},
48 silly batched     localDependents: Map {} } ]
49 info publish Publishing packages to npm...
50 notice Skipping all user and access validation due to third-party registry
51 notice Make sure you're authenticated properly ¯\_(ツ)_/¯
52 warn ENOLICENSE Packages @lerna-repo/repo2 are missing a license
53 silly getCurrentSHA
54 verbose getCurrentSHA 394d73d06d13a133729d1704141997a8e069706c
55 silly lifecycle No script for "prepublish" in "root", continuing
56 silly lifecycle No script for "prepare" in "root", continuing
57 silly lifecycle No script for "prepublishOnly" in "root", continuing
58 silly lifecycle No script for "prepack" in "root", continuing
59 verbose pack-directory packages\repo2
60 silly lifecycle No script for "prepublish" in "@lerna-repo/repo2", continuing
61 silly lifecycle No script for "prepare" in "@lerna-repo/repo2", continuing
62 silly lifecycle No script for "prepublishOnly" in "@lerna-repo/repo2", continuing
63 silly lifecycle No script for "prepack" in "@lerna-repo/repo2", continuing
64 silly lifecycle No script for "postpack" in "@lerna-repo/repo2", continuing
65 verbose packed @lerna-repo/repo2 packages\repo2
66 silly lifecycle No script for "postpack" in "root", continuing
67 verbose publish @lerna-repo/repo2
68 http fetch PUT 404 http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2 258ms
69 error Error: 404 Not Found - PUT http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2
69 error     at res.buffer.catch.then.body (D:\lerna-repo\node_modules\npm-registry-fetch\check-response.js:104:15)
```
</p></details>

My `package.json` scripts is
```json
  "scripts": {
    "deploy": "lerna publish --yes",
    "echo": "echo next step",
    "example": "yarn deploy && yarn echo"
  }
```
When I run script `yarn example`, if the 1st sub-script `yarn deply` failed, `yarn echo` should not be executed, but it was executed indeed.
<details><summary>command-window-execute-result.txt</summary><p>


```text
PS D:\lerna-repo> yarn example
yarn run v1.12.3
$ yarn deploy && yarn echo
$ lerna publish --yes
lerna notice cli v3.10.5
lerna info versioning independent
lerna info Looking for changed packages since @lerna-repo/repo2@2.1.1

Changes:
 - @luban-lerna-repo/repo2: 2.1.2 => 2.1.3

lerna info auto-confirmed
lerna info git Pushing tags...
lerna info publish Publishing packages to npm...
lerna notice Skipping all user and access validation due to third-party registry
lerna notice Make sure you're authenticated properly ¯\_(ツ)_/¯
lerna WARN ENOLICENSE Packages @lerna-repo/repo2 are missing a license
lerna http fetch PUT 404 http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2 258ms
lerna ERR! Error: 404 Not Found - PUT http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2
lerna ERR!     at res.buffer.catch.then.body (D:\lerna-repo\node_modules\npm-registry-fetch\check-response.js:104:15)
lerna ERR! lerna 404 Not Found - PUT http://internal.company.com/npm-repo/npmjs/@lerna-repo%2frepo2
$ echo next step
next step
Done in 14.41s.
``` 
</p></details>

**Comment by evocateur** [comment]
Ah yes, much clearer, thanks!

Definitely missing some code in the catch handlers (or `catch` entirely) around the [publish loop](https://github.com/lerna/lerna/blob/master/commands/publish/index.js#L583-L599). It occurs to me that it should probably be handling the error directly in the utility that [calls libnpmpublish](https://github.com/lerna/lerna/blob/c79a82722095607940749f5c6fa1cdd90c56a437/utils/npm-publish/npm-publish.js#L53):
```patch
-      return publish(manifest, tarData, opts);
+      return publish(manifest, tarData, opts).catch(err => {
+        // ensure process exits non-zero
+        process.exitCode = err.errno || 1;
+
+        // re-throw to break chain upstream
+        throw err;
+      });
```

(dry-coded, definitely needs tests)

**Comment by evocateur** [comment]
https://github.com/lerna/lerna/commit/9e9ce0859f03636ed5f36a5e619a60de5b4c938b fixed this, forgot to tag the bug

**Comment by evocateur** [comment]
published in [v3.10.7](https://github.com/lerna/lerna/releases/tag/v3.10.7)
